src/basic/hashmap.c

```diff
@@ -2149,6 +2151,40 @@ int _hashmap_dump_sorted(HashmapBase *h, void ***ret, size_t *ret_n) {
 
         typesafe_qsort_r(entries, n, hashmap_entry_compare, h->hash_ops->compare);
 
+        *ret = TAKE_PTR(entries);
+        *ret_n = n;
+        return 0;
+}
+
+int _hashmap_dump_keys_sorted(HashmapBase *h, void ***ret, size_t *ret_n) {
+        _cleanup_free_ struct hashmap_base_entry **entries = NULL;
+        size_t n;
+        int r;
+
+        r = _hashmap_dump_entries_sorted(h, &entries, &n);
+        if (r < 0)
+                return r;
+
+        /* Reuse the array. */
+        FOREACH_ARRAY(e, entries, n)
+                *e = (void*) (*e)->key;
```

Comment:
uh, that's mean. this as an array of a specific type now, and now you assign arbitrary stuff into it.

That's not just ugly, but also wrong, since the compiler can assume that pointers are always aligned to the alignment of their type. but the key field here might not be, yet you assign it to a `struct hashmap_base_entry*` pointer, which implies alignment.

Reply:
what you are doing here would probably be OK, if you'd change the type of `entries` to `void**`, i.e. don't claim anything about any type. would then have to cast this when calling _hashmap_dump_entries_sorted i guess

Reply:
This is how it was before just only with values. Perhaps the whole thing just needs a bit of a rework...

> since the compiler can assume that pointers are always aligned to the alignment of their type

I'm not quite sure I'm following this here. Pointers are always 8 bytes, and we're allocating the array via malloc. Arrays are always tightly packed in C, no? So what room does the compiler have to care about alignment?

Reply:
Opened https://github.com/systemd/systemd/pull/31302, let's continue there